### PR TITLE
Correction of Generic Contaminant Handling in Some Situations

### LINKED ISSUE #10473: Make sure that CO2 and GenContam are handled by all components
Issue overview
--------------
1. While reviewing another PR, noticed that `GenContam` isn't getting passed through by variable speed coils, only `CO2` is.
https://github.com/NREL/EnergyPlus/blob/378844d64bd81314354a7f30ab5fb3a9c2afa18d/src/EnergyPlus/VariableSpeedCoils.cc#L8358-L8360

2. While working on #10348 noticed this for SpaceHVAC:ZoneEqupimentMixer.
https://github.com/NREL/EnergyPlus/blob/378844d64bd81314354a7f30ab5fb3a9c2afa18d/src/EnergyPlus/DataZoneEquipment.cc#L1839-L1850

3. Need to check the entire code base to make sure that `CO2` and `GenContam` are being calculated and/or passed through as needed.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by RKStrand**
Pull request overview
---------------------
 - Fixes #10473 
 - While development team members were working on other defects, it was noticed that there were a couple of places in the code where the generic contaminant information was not being handled properly, i.e. it wasn't being updated properly.  This pull request fixes this issue and now all code should be handling generic contaminant updating properly.  As part of this fix, a thorough check of the entire code was made to ensure that both CO2 and generic contaminant code was present in all locations.  During that check, an unnecessary/misplaced line of code was found and eliminated as well.  While no documentation fixes were need as part of this work, a new unit test was created and an existing unit test was enhanced to test this correction.  No changes in the output were noted and the unit tests show that updating is being handled correctly now.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/DataZoneEquipment.cc [hunk 1]
```diff
@@ -1810,12 +1811,14 @@ void ZoneEquipmentMixer::setOutletConditions(EnergyPlusData &state)
         sumEnthalpy += spaceOutletNode.Enthalpy * mixerSpace.fraction;
         sumHumRat += spaceOutletNode.HumRat * mixerSpace.fraction;
         sumCO2 += spaceOutletNode.CO2 * mixerSpace.fraction;
+        sumGenContam += spaceOutletNode.GenContam * mixerSpace.fraction;
```
Comment (mjwitte) [inline src/EnergyPlus/DataZoneEquipment.cc]: Most (all?) other places that work with CO2 and GenContam have ifs like `if (state.dataContaminantBalance->Contaminant.GenericContamSimulation) {`. Should there be ifs here?

File: src/EnergyPlus/VariableSpeedCoils.cc [hunk 2]
```diff
@@ -8337,6 +8337,10 @@ namespace VariableSpeedCoils {
             state.dataLoopNodes->Node(AirOutletNode).CO2 = state.dataLoopNodes->Node(AirInletNode).CO2;
         }
 
+        if (state.dataContaminantBalance->Contaminant.GenericContamSimulation) {
+            state.dataLoopNodes->Node(AirOutletNode).GenContam = state.dataLoopNodes->Node(AirInletNode).GenContam;
+        }
```
Comment (mjwitte) [inline src/EnergyPlus/VariableSpeedCoils.cc:8342]: @Myoldmopar There are ifs like this scattered throughout the code. For simple assignments like this are these ifs saving any time or are they just clutter?

File: src/EnergyPlus/DataZoneEquipment.cc (hunk 1 above)
Comment (RKStrand) [reply src/EnergyPlus/DataZoneEquipment.cc]: @mjwitte hmmm…that’s a good point. I will look at implementing such if checks when I get back to this next week on Tuesday.

Comment (RKStrand) [reply src/EnergyPlus/DataZoneEquipment.cc]: @mjwitte Ok, I've added the IF checks as per your suggestion and just now made a commit (as well as merged in develop to get everything back up to speed).  Anticipating that this should come back all green as before.

**Review by mjwitte — APPROVED** [review]
Code changes look good. Unit tests pass. CI is clean and green. Merging. Thanks @RKStrand !

File: src/EnergyPlus/VariableSpeedCoils.cc (hunk 2 above)
Comment (rraustad) [reply src/EnergyPlus/VariableSpeedCoils.cc:8342]: I agree these checks aren't needed (just pass the 0). Also thinking of a safeAirNodeCopy function passing MassFlowRate, Quality, Press, MassFlowRateMin, MassFlowRateMax, MassFlowRateMinAvail and MassFlowRateMaxAvail, and CO2 and GenContam. A lot of these child update functions would reduce to:

        safeAirNodeCopy(state, outletnode, inletnode);
        state.dataLoopNodes->Node(AirOutletNode).Temp = varSpeedCoil.OutletAirDBTemp;
        state.dataLoopNodes->Node(AirOutletNode).HumRat = varSpeedCoil.OutletAirHumRat;
        state.dataLoopNodes->Node(AirOutletNode).Enthalpy = varSpeedCoil.OutletAirEnthalpy;

**Pull request merged**